redash/handlers/dashboards.py

```diff
@@ -107,6 +107,10 @@ def post(self):
         )
         models.db.session.add(dashboard)
         models.db.session.commit()
+
+        dashboard.record_changes(self.current_user, models.Change.Type.Created)
+        models.db.session.commit()
```

Comment:
Why commit twice? If it's because the object doesn't exist for the reference in changes -- SQLA knows to take care of this and supposed to save them in order and update the id. Did you try it?

Reply:
Tried, and without this `commit()` I get this error: `sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) null value in column "object_id" violates not-null constraint`. So I guess it doesn't work (maybe in some cases SQLA can handle this, but in this particular case it didn't)

Reply:
It makes sense, but it will be harder to add change tracking to new models (e.g. Alerts). I think properly configured indexes should work fine as well

Reply:
> Maybe because we're using non "traditional" foreign key it fails.

I've read some docs about SQLA and how it handles data changes, so I guess I know why it happens. When instance it created - it lives on its own until added to any session. When `sesson.commit()` is called - SQLA takes all new/dirty/removed instanced associated with the session, creates a SQL query for each of them, executes and only after that updates instances. In a case with new instances that means that `id` will be populated only _after_ `session.commit()`. I think it has some special handling for a case when there is a properly set-up relation between models, but `GFKBase` does not have relations in SQLA's meaning, therefore SQLA doesn't know that `Change` instances depend on `Query`/`Dashboard`/etc. being saved within the same `session.commit()`